dev/mail#90 - Allow re-use mailing on draft mailings and adhoc mailings

larssandergreen opened the pull request:
See also : https://lab.civicrm.org/dev/mail/-/issues/90

## Before

* "Re-use" was shown only on scheduled/sent and archived mailings pages.
* "Re-use" was shown only on mailings with standard groups (omitting maliings that used adhoc targets).

## After

* "Re-use" is also shown on draft/unscheduled mailings (i.e. on all mailings).
* "Re-use" is also shown on mailings with adhoc targets.
* The list of actions has been reordered, so that "Continue" comes before "Re-use"/"Copy".
    ![image](https://user-images.githubusercontent.com/25517556/114755428-d9920980-9d16-11eb-9562-d7e0a42c9d1b.png)
    ![image](https://user-images.githubusercontent.com/25517556/114755442-dd259080-9d16-11eb-8a6b-9928288aeac8.png)
* The action "Re-use" has been renamed to "Copy".

NOTE: Description updated April 19 to indicate that  the PR also affects adhoc mailings.

larssandergreen committed:
Allow re-use mailing on draft mailings

Before, re-use was shown only on scheduled/sent and archived mailings pages, now shown on draft/unscheduled as well (i.e. on all mailings pages).
Also re-ordered options to put Re-use after Continue.

civicrm-builder commented:
Can one of the admins verify this patch?

eileenmcnaughton commented:
add to whitelist

eileenmcnaughton commented on CRM/Mailing/Selector/Browse.php:
```diff
@@ -409,6 +403,10 @@ public function &getRows($action, $offset, $rowCount, $sort, $output = NULL) {
 
+        if ($allAccess || $showCreateLinks) {
+          $actionMask |= CRM_Core_Action::UPDATE;
+          }
```
this indentation is going to cause jenkins to scream blue murder - tests are running now so it looks like

![image](https://user-images.githubusercontent.com/336308/114806194-6eb8ff00-9df8-11eb-8283-2bee6d105434.png)


But I'm betting the brown circle turns into a red x - you can click through on the x to see what jenkins has to say

larssandergreen replied:
fixed

larssandergreen committed:
Fixed bad bracket indent

totten commented:
On a functionality level, I think this is a good idea. 👍 

<!-- CiviCRM Review Template WORD-1.2 from https://raw.githubusercontent.com/civicrm/civicrm-dev-docs/master/docs/standards/review/template-word-1.0.md -->

<!-- In each category, change the word "Undecided" to "Pass" or "Issue". Add explanatory comments if prompted or desired. -->

* General standards
    * ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) __Pass__
    * ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __Discuss__:
        * I was looking at two buttons ("Continue", "Re-use") and clicked "Re-use". After clicking, I wasn't really sure that it had made a copy. (Basically, the UX of clicking "Continue" or "Re-use" is exactly the same.) I had to put on my developer-hat, look at the URL to see the mailing ID#, and reassure myself. It would be nice to have some cue -- like changing the `mailing.name` ("Copy of {$oldName}") or showing a status-alert (`CRM_Core_Session::setStatus(...ts('Created a new copy of mailing "%1"', ...)...)`).
    * ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __Pass__
    * ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __Pass__
* Developer standards
    * ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Pass__
    * ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Pass__: Looks pretty clean and nicely matches other code in the same context.
    * ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Pass__
    * ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __Pass__

totten commented on CRM/Mailing/Selector/Browse.php:
```diff
@@ -264,6 +258,12 @@ public function &getRows($action, $offset, $rowCount, $sort, $output = NULL) {
           'title' => ts('Continue Mailing'),
         ],
+        CRM_Core_Action::UPDATE => [
+          'name' => ts('Re-Use'),
```
How hard would it be to make this label adapt?

This is maybe a bit nitpicky, but "Re-Use" suggests that the mailing has been "used" before. A draft hasn't been used  before.

If it's not too hard, maybe something like this would be more apropos:

```php
  'name' => $thisItemIsADraft ? ts('Copy') : ts('Re-Use'),
```

larssandergreen replied:
Agreed, I suggested [in the issue](https://lab.civicrm.org/dev/mail/-/issues/90) changing the text to "Copy". My thought is that it would be better to keep the label consistent for all mailings and just change it to "Copy" throughout. Otherwise, users may wonder what the difference is between "Copy" and "Re-use". Copy is also consistent with events, contribution pages, price sets, etc which all use some variation of Copy, Make a Copy, Copy Price Set, etc.

totten replied:
Yup. "Copy" is a small UI change in near term (*and we might want to tweak the docs after this PR*), but I think you're right - "Copy" is more consistent in the big picture. It also matches the `Copy of ...` text.

magnolia61 commented:
love this! before you had to sent a mailing in order to reuse it. Nice one :-)

mlutfy commented:
+1 on this, I often had to use the hidden URLs to do this.

Warning to those using multi-lingual and using this process to translate mailings: https://lab.civicrm.org/dev/mail/-/issues/87

larssandergreen commented:
@totten Yes, good idea to give the user some feedback to indicate this is a copy. I think adding "- Copy" or something like that to the mailing name makes the most sense and is consistent with how this works for events, price sets, contribution pages, etc.
But how the mailing copy works is a bit beyond me right now, so someone would need to point me in the right direction. I see how this works for price sets, for example, but I don't see anything analogous for mailings.

totten commented:
> But how the mailing copy works is a bit beyond me right now, so someone would need to point me in the right direction.

Good question. It can be tricky to trace the actions for some of these forms. Here's how one can locate this one:

* The link declared `'url' => 'civicrm/mailing/send',`, so that's our focal point.
* Traditionally, I would've just done a global text search for `civicrm/mailing/send`. Most IDEs have some hotkey for this, but I'm a CLI fan, so I'd do `grep`:

    ```
    $ grep -r civicrm/mailing/send CRM/
    CRM/Core/Block.php:          'path' => 'civicrm/mailing/send',
    CRM/Utils/Token.php:        $value = CRM_Utils_System::url('civicrm/mailing/send',
    CRM/Mailing/Selector/Browse.php:          'url' => 'civicrm/mailing/send',
    CRM/Mailing/Selector/Browse.php:          'url' => 'civicrm/mailing/send',
    CRM/Mailing/xml/Menu/Mailing.xml:    <path>civicrm/mailing/send</path>
    CRM/Mailing/Controller/Send.php:    // New:            civicrm/mailing/send?reset=1
    CRM/Mailing/Controller/Send.php:    // Re-use:         civicrm/mailing/send?reset=1&mid=%%mid%%
    CRM/Mailing/Controller/Send.php:    // Continue:       civicrm/mailing/send?reset=1&mid=%%mid%%&continue=true
    ```

    The key result is the ["Menu" XML file](https://docs.civicrm.org/dev/en/latest/framework/routing/), which declares that `<path>civicrm/mailing/send</path>` corresponds to `<page_callback>CRM_Mailing_Controller_Send</page_callback>`.

* Alternatively, instead of a text-search, now-a-days I'll do:
    ```
    $ cv api4 Route.get +w path=civicrm/mailing/send
    [
        {
            "path": "civicrm/mailing/send",
            "title": "New Mailing",
            "page_callback": "CRM_Mailing_Controller_Send",
    ...
    ```

* Either way, we land at [CRM_Mailing_Controller_Send](https://github.com/civicrm/civicrm-core/blob/5.37/CRM/Mailing/Controller/Send.php). The key bit appears to be:

    https://github.com/civicrm/civicrm-core/blob/bc77d7c0798081f999e6eca92d75fa874fd04e74/CRM/Mailing/Controller/Send.php#L50-L53

The easiest thing is to probably to keep the call to `civicrm_api3('Mailing', 'clone', ...)`. Afterward, write the new `name`, call `CRM_Core_Session::setStatus()`, and then `redirect()` at the end.

larssandergreen committed:
Add " - Copy" to the end of the new mailing name when cloning

larssandergreen committed:
Changed " - Copy" append to "Copy of " prepend for consistency

larssandergreen commented:
Thanks @totten, that helped a lot. Now adds "Copy of" to the cloned mailing.

What are your thoughts on changing to "Copy" instead of "Re-use" on all mailing pages (see note above)?

totten commented on CRM/Mailing/Controller/Send.php:
```diff
@@ -49,6 +49,10 @@ public function __construct($title = NULL, $action = CRM_Core_Action::NONE, $mod
       $clone = civicrm_api3('Mailing', 'clone', ['id' => $mid]);
+      civicrm_api3('Mailing', 'create', [
+        'id' => $clone['id'],
+        'name' => "Copy of " . $clone['values'][$clone['id']]['name'],
```
Yup, I think it's better with `Copy of...` instead of `... - Copy`. Matches some similar bits elsewhere.

Actually, borrowing from `CRM/SMS/Form/Group.php` (which also has a `Copy of...`), we should use `ts()`. (Prose should usually be translated.)

```suggestion
        'name' => ts('Copy of %1', [1 => $clone['values'][$clone['id']]['name']]) ,
```

totten committed:
Update CRM/Mailing/Controller/Send.php

totten commented:
Re-tested and still looks good. If you can update the label to "Copy", that'll be perfect. Flagging "merge ready" (ie *it safe to merge now, but we can wait a day or so in case you push the update*).

larssandergreen committed:
Change Re-use to Copy

larssandergreen commented:
Changed Re-use to Copy here and added a [documentation pull request for the same.](https://lab.civicrm.org/documentation/docs/user-en/-/merge_requests/479).

larssandergreen commented:
Test failure is a different contact checksum from a token.

larssandergreen commented:
jenkins, test this please

larssandergreen commented:
[@ufundo has helped me figure](https://lab.civicrm.org/dev/mail/-/issues/90#note_58155) out that this patch does something else as well: Before, mailings created from searches could not be re-used/copied. Now, they can be copied, the same as any other mailing.
That's nice, but I just want to check if there is any reason we should be preventing mailings from search results from being copied. Is the concern that the hidden smart group will no longer be present?

totten commented:
Yeah, that's a bit curious - why would it block Copy/Re-use of search-based mailings (`CRM_Mailing_Form_Task_AdhocMailing`; aka mailings with hidden groups; aka mailings where the targets are `group_type=Base`)?

I agree with @ufundo @larssandergreen that this seems like an improvement. It is a fairly distinct change, so I've updated the `r-explain`.

I'm not sure why it was hidden, but - as an exercise in due diligence - I can imagine two plausible reasons. (Neither seems particularly compelling to me right now.) They are:

1. Adhoc mailings require special bits in the editor UI (e.g. handling of the "Recipients" field and the "Unsubscribe" field). This has evolved over time, and it's fairly isolated in Civi 5.x (*it's only two fields that change*), and these two fields seem to be working. But circa Civi 2.x/3.x (*when this was added*), the whole screen worked differently, and it may have been buggy if you tried to re-use an adhoc mailing.
2. If you think of the mailing as "ad hoc", then the selection of contacts is a fleeting decision of the moment.  The list of recipients that you chose on Tuesday may no longer be correct for a future mailing on Friday.
    *  (a) Example: You search for event participants and cherry-pick a few who you know to be donors. Then you send the first message. Two weeks later, you copy/reuse and send another message. But in the intervening time, the various statuses (registered participants; existing donors) may have changed. If you re-use the mailing, then you're acting with stale list of contacts.
    * (b) Counter-example: You had an informal chat with a group of people after a meetup. You decide to send those people a follow-up by email. Two weeks later, you send them another follow-up message. In this case, you might re-use the mailing to avoid needing to re-select people.
    * Opinion: I think (2b) is legit enough. It's true that (2a) could be mis-use, but even with (2a)... there may be legit reasons to re-use the original target-list. *It really depends on what/why you're communicating.* The main thing is to communicate clearly so that the user makes an informed decision about copying/reusing values.

Aside: When you actually re-use a mailing, you see this "Hidden group NNN", which is mysterious. But you can inspect/drill-down on the full "recipients" list, so I don't think it's a hard block. (Maybe a label like "Adhoc group NNN" would be a little more apropos... but that's getting really outside the scope of the issue.)

Anyway, this still looks fine. Merging.

larssandergreen commented:
Great, thanks @totten. Regarding the hidden smart group aside, I actually [have an issue open about this](https://lab.civicrm.org/dev/core/-/issues/2521#note_57583). With help from @tschuettler and some testing and digging through the code , it looks like the the hidden group can in fact be smart or not smart, so you could actually make a hidden smart group which would presumably update for the next mailing (but I certainly could be wrong about the next mailing case in particular).
For some reason, the hidden smart group isn't updated between when you create the mailing and when you send it or come back to it later - unless you clear the smart group cache manually or via cron. Normal smart groups don't behave that way. I don't know why these are different and don't really know the first place to start with this one. But if you have thoughts, that might help clarify what is happening here.
